gms/gossiper.cc

```diff
@@ -320,29 +320,75 @@ future<> gossiper::handle_ack_msg(msg_addr id, gossip_digest_ack ack_msg) {
             return make_ready_future<>();
         }
-        /* Get the state required to send to this gossipee - construct GossipDigestAck2Message */
-        std::map<inet_address, endpoint_state> delta_ep_state_map;
-        for (auto g_digest : g_digest_list) {
-            inet_address addr = g_digest.get_endpoint();
-            auto local_ep_state_ptr = this->get_state_for_version_bigger_than(addr, g_digest.get_max_version());
-            if (local_ep_state_ptr) {
-                delta_ep_state_map.emplace(addr, *local_ep_state_ptr);
-            }
+        ack_msg_pending& p = _ack_handlers[from.addr];
+        if (p.pending) {
+            // The latest ack message digests from peer has the latest infomation, so
+            // it is safe to drop the previous ack message digests and keep the latest
+            // one only.
+            logger.debug("Queue gossip ack msg digests from node {}, ack_msg_digest={}", from, ack_msg_digest);
+            p.ack_msg_digest = std::move(ack_msg_digest);
+            return make_ready_future<>();
+        } else {
+            // Process the ack message immediately
+            logger.debug("Process gossip ack msg digests from node {}, ack_msg_digest={}", from, ack_msg_digest);
+            p.pending = true;
+            return do_with(std::move(ack_msg_digest), [this, from] (utils::chunked_vector<gossip_digest>& ack_msg_digest) mutable {
+                return repeat([this, from, &ack_msg_digest] {
+                    return do_send_ack2_msg(from, std::move(ack_msg_digest)).then([this, from, &ack_msg_digest] () mutable {
+                        if (!_ack_handlers.count(from.addr)) {
+                            return stop_iteration::yes;
+                        }
+                        ack_msg_pending& p = _ack_handlers[from.addr];
+                        if (p.ack_msg_digest) {
+                            // Process pending gossip ack msg digests and send ack2 msg back
+                            logger.debug("Handle queued gossip ack msg digests from node {}, ack_msg_digest={}, pending={}",
+                                    from, p.ack_msg_digest, p.pending);
+                            ack_msg_digest = std::move(p.ack_msg_digest.value());
+                            p.ack_msg_digest= {};
+                            return stop_iteration::no;
+                        } else {
+                            // No more pending ack msg digests to process
+                            p.pending = false;
+                            logger.debug("No more queued gossip ack msg digests from node {}, ack_msg_digest={}, pending={}",
+                                    from, p.ack_msg_digest, p.pending);
+                            return stop_iteration::yes;
+                        }
+                    }).handle_exception([this, from] (std::exception_ptr ep) {
+                        if (_ack_handlers.count(from.addr)) {
```

Comment:
I don't think that `_ack_handlers` handling has to be different after failure than after success. You could turn the `then()` case into `finally()` case and drop this code.

Reply:
The exception handler makes sure the we clear the p.pending in case of error. Otherwise, we could queue the syn msg while the loop keeps handling the syn is not running. 

IMO, it is more clear to handle the normal path and the error path.

Reply:
> You could have:
> 
> ```
> repeat([] {
>    return send().handle_exception([] {
>       // log
>    }).then([] {
>          // maintain pending
>    });
> });
> ```

if send() throws immediately, the handle_exception won't be called as neither. 

I think the fix would be using futurize_apply inside send.

Reply:
> Would be nice to make it generic to avoid duplication in the other patch.

They looks similar but the msg to send, the log message and the pending msg to store are different. I want to avoid a template function or a helper function.

Reply:
Why do you want to avoid a template function?

Reply:
> Because the gain is small while it complicates the code and makes compile time longer.

Why would it complicate the code? I think it simplifies it, instead of having the algorithm rolled out independently in two functions, we have it only in one. So the complexity cuts in half.

The algorithm is not trivial, we've found bugs in it.

Reply:
It complicates the code in the sense we need to make the code generic to support both cases, e.g.,  call do_send_ack_msg or do_send_ack2_msg, check p.syn_msg or p.ack_msg_digest, log different things. 

It does not cut the complexity in half, because they are very similar. If we find one bug in one place, we can fix the other one very easily. 

That being said, this is an urgent user issue. Let's not debate on small things like if we should introduce a template function to reduce a few lines code.

Reply:
> It complicates the code in the sense we need to make the code generic to support both cases, e.g., call do_send_ack_msg or do_send_ack2_msg, check p.syn_msg or p.ack_msg_digest, log different things.

You're probably thinking about a different meaning of "complex". 

Making this code generic reduces complexity, even if we wouldn't have the second user, because it abstracts away the details that are irrelevant to achieve the task so that the number of relationships that need to be considered is as small as possible. Without this abstraction, when you look at the code, you have no idea what is the goal behind it. You have to go and read the call to `do_send_ack_msg()` to convince yourself that it doesn't touch `p.syn_msg` and doesn't affect the outer control flow, and what we're really achieving is to limit the concurrency of that call. 

What we have here is a kind of `semaphore` with the concurrency of 1 and waitlist limited to just the last task. If that kind of synchronizer was made a separate class, like a `semaphore` is, all this would be immediately obvious.

Another advantage of abstracting it is that you could unit-test this separately from the gossiper. Currently, when it's woven into the gossiper code like this it's hard to test the corner cases.

> 
> It does not cut the complexity in half, because they are very similar. If we find one bug in one place, we can fix the other one very easily.

There's a high chance that whoever finds a bug in one place won't be aware that the same bug exists in another one.

> 
> That being said, this is an urgent user issue.

I was told that it's not more urgent than other fixes.

Reply:
> > It complicates the code in the sense we need to make the code generic to support both cases, e.g., call do_send_ack_msg or do_send_ack2_msg, check p.syn_msg or p.ack_msg_digest, log different things.
> 
> You're probably thinking about a different meaning of "complex".
> 
> Making this code generic reduces complexity, even if we wouldn't have the second user, because it abstracts away the details that are irrelevant to achieve the task so that the number of relationships that need to be considered is as small as possible. Without this abstraction, when you look at the code, you have no idea what is the goal behind it. You have to go and read the call to `do_send_ack_msg()` to convince yourself that it doesn't touch `p.syn_msg` and doesn't affect the outer control flow, and what we're really achieving is to limit the concurrency of that call.
> 
> What we have here is a kind of `semaphore` with the concurrency of 1 and waitlist limited to just the last task. If that kind of synchronizer was made a separate class, like a `semaphore` is, all this would be immediately obvious.

You are welcome to send patch to do such abstraction. The flow is pretty gossip specific. And I doubt such "semaphore with the concurrency of 1 and waitlist limited to just the last task" will be useful for other users. 

> 
> Another advantage of abstracting it is that you could unit-test this separately from the gossiper. Currently, when it's woven into the gossiper code like this it's hard to test the corner cases.

Even if you abstract the flow, it will still be pretty hard to do the unit-test because of exchange of rpc messages.  

> > It does not cut the complexity in half, because they are very similar. If we find one bug in one place, we can fix the other one very easily.
>
> There's a high chance that whoever finds a bug in one place won't be aware that the same bug exists in another one.

This argument is true but there is even a higher chance the one fixes the bug is me who is aware the situation. 

Abstraction is nice but let's not waste more time on debating it. We can create an issue for the abstraction and do it later if we think it is really important.

> 
> > That being said, this is an urgent user issue.
> 
> I was told that it's not more urgent than other fixes.

It is less urgent than other fixes in your queue does not mean this issue is not urgent. Losing a user is very urgent to me. Using template to reduce a few lines of code is not urgent to me.

Reply:
> > > It complicates the code in the sense we need to make the code generic to support both cases, e.g., call do_send_ack_msg or do_send_ack2_msg, check p.syn_msg or p.ack_msg_digest, log different things.
> > 
> > 
> > You're probably thinking about a different meaning of "complex".
> > Making this code generic reduces complexity, even if we wouldn't have the second user, because it abstracts away the details that are irrelevant to achieve the task so that the number of relationships that need to be considered is as small as possible. Without this abstraction, when you look at the code, you have no idea what is the goal behind it. You have to go and read the call to `do_send_ack_msg()` to convince yourself that it doesn't touch `p.syn_msg` and doesn't affect the outer control flow, and what we're really achieving is to limit the concurrency of that call.
> > What we have here is a kind of `semaphore` with the concurrency of 1 and waitlist limited to just the last task. If that kind of synchronizer was made a separate class, like a `semaphore` is, all this would be immediately obvious.
> 
> You are welcome to send patch to do such abstraction. The flow is pretty gossip specific. And I doubt such "semaphore with the concurrency of 1 and waitlist limited to just the last task" will be useful for other users.

Usefulness for other users is orthogonal. Reducing complexity within the gossiper is still worth it.

We already have `serialized_action` which does something very similar. It's used to limit the concurrency of schema syncs.

> 
> > Another advantage of abstracting it is that you could unit-test this separately from the gossiper. Currently, when it's woven into the gossiper code like this it's hard to test the corner cases.
> 
> Even if you abstract the flow, it will still be pretty hard to do the unit-test because of exchange of rpc messages.

The abstraction would not do any RPC calls, it would work with async actions and would be externally triggered like `serialized_action`.

> 
> > > It does not cut the complexity in half, because they are very similar. If we find one bug in one place, we can fix the other one very easily.
> > 
> > 
> > There's a high chance that whoever finds a bug in one place won't be aware that the same bug exists in another one.
> 
> This argument is true but there is even a higher chance the one fixes the bug is me who is aware the situation.
> 
> Abstraction is nice but let's not waste more time on debating it. We can create an issue for the abstraction and do it later if we think it is really important.
> 
> > > That being said, this is an urgent user issue.
> > 
> > 
> > I was told that it's not more urgent than other fixes.
> 
> It is less urgent than other fixes in your queue does not mean this issue is not urgent. Losing a user is very urgent to me.

I didn't mean that it's less urgent relative to other things in my queue, but that it's not more urgent than other fixes which are regularly submitted. So it's not as urgent so that we should sacrifice code quality and merge in a hurry. Also, not urgent does not mean that it's not important. 

I can see that Shlomi pushes for this to be merged, so maybe it is more urgent than I thought. I can merge this as is, but I really think this deserves a cleanup.

> Using template to reduce a few lines of code is not urgent to me.